src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs

```diff
@@ -0,0 +1,4051 @@
+                public static CancellationTokenSource CreateLinkedTokenSource(CancellationToken token1, CancellationToken token2);
+                public static CancellationTokenSource CreateLinkedTokenSource(params CancellationToken[] tokens);
+            }
+            */
+            return CS8VerifyAnalyzerAsync(@"
+using System.Threading;
+class C
+{
+    void M(CancellationToken ct)
+    {
+        CTS.Method(ct); // Don't diagnose
+    }
+}
+class CTS
+{
+    public static void Method(CancellationToken token){}
+    public static void Method(params CancellationToken[] tokens){}
+    public static void Method(CancellationToken token1, CancellationToken token2){}
+}
+            ");
+        }
+
+        [Fact]
+        public Task CS_NoDiagnostic_CancellationTokenSource_ParamsUsed_Order3()
+        {
+            return CS8VerifyAnalyzerAsync(@"
+using System.Threading;
+class C
+{
+    void M(CancellationToken ct)
+    {
+        CTS.Method(ct); // Don't diagnose
+    }
+}
+class CTS
+{
+    public static void Method(CancellationToken token1, CancellationToken token2){}
+    public static void Method(params CancellationToken[] tokens){}
+    public static void Method(CancellationToken token){}
+}
+            ");
+        }
+
+        [Fact]
+        public Task CS_NoDiagnostic_CancellationTokenSource_ParamsUsed_Order4()
+        {
+            return CS8VerifyAnalyzerAsync(@"
+using System.Threading;
+class C
+{
+    void M(CancellationToken ct)
+    {
+        CTS.Method(ct); // Don't diagnose
+    }
+}
+class CTS
+{
+    public static void Method(CancellationToken token1, CancellationToken token2){}
+    public static void Method(CancellationToken token){}
+    public static void Method(params CancellationToken[] tokens){}
+}
+            ");
+        }
+
+        [Fact]
+        public Task CS_NoDiagnostic_CancellationTokenSource_ParamsUsed_Order5()
+        {
+            return CS8VerifyAnalyzerAsync(@"
+using System.Threading;
+class C
+{
+    void M(CancellationToken ct)
+    {
+        CTS.Method(ct); // Don't diagnose
+    }
+}
+class CTS
+{
+    public static void Method(params CancellationToken[] tokens){}
+    public static void Method(CancellationToken token){}
+    public static void Method(CancellationToken token1, CancellationToken token2){}
+}
+            ");
+        }
+
+        [Fact]
+        public Task CS_NoDiagnostic_CancellationTokenSource_ParamsUsed_Order6()
+        {
+            return CS8VerifyAnalyzerAsync(@"
+using System.Threading;
+class C
+{
+    void M(CancellationToken ct)
+    {
+        CTS.Method(ct); // Don't diagnose
+    }
+}
+class CTS
+{
+    public static void Method(params CancellationToken[] tokens){}
+    public static void Method(CancellationToken token1, CancellationToken token2){}
+    public static void Method(CancellationToken token){}
+}
+            ");
+        }
+
+        [Fact]
+        public Task CS_NoDiagnostic_ExtensionMethodTakesToken()
```

Comment:
@mavasani I built the dotnet/runtime repo to detect this rule, and I found that the analyzer is not behaving the same way as in the unit tests:
- In this unit test, which passes, we do not offer a diagnostic when the overload method is located in an extension method.
- In dotnet/runtime, the extension method is found, a diagnostic is offered, as well as a fix, and applying the fix causes the object to get substituted with the extension class name, and suddenly the code does not build.

I'll tag you in the runtime PR to show you the exact example.

What could be causing this?

Reply:
PR: https://github.com/dotnet/runtime/pull/37607

Reply:
Wondering if this is caused by presence of using directive that brings in Extensions into the file or not? Regardless, I think you should add a check in the analyzer to ignore all candidate overloads with [IsExtensionMethod](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Symbols/IMethodSymbol.cs,42) property set to true.

Reply:
I found the root cause, managed to repro it in my unit tests, and addressed this case. Now the expression of extensions methods will remain untouched. The only thing that I couldn't figure out how to address was that the added cancellation token argument is getting added with an explicit name, because I couldn't find a clear way to recognize the first "this" argument (the `.IsThis` property is showing up as false, seems like a bug).